README.md
# github-actions  
## General information
Test repo to play around with github actions  

/.github/workflows contains the currently used workflows  
/testing_workflows contains workflows that are currently beeing evaluated, tested, or kept to play around  

## Explanation for every workflow.yml file  

### label_pull_request.yml (Label Pull Request)

Whenever a pull request is opened, edited, synchronized, reopened  
this workflow gets triggered.  

It contains the following **Actions**:  
* Label pull request
  * Currently just echos something for testing
  * Could be used to add additional labels to pull requests (e.g. if review approved add ready for approval/merge label)
* Label pull request for RC
  * Currently uses TimonVS/pr-labeler-action with the "pr-labeler.yml" file defining the labels  
    and their matching according to our RC branch naming convention
  * Only run if merge target is master
  * Possible label values:
    * Leading to release: patch, minor, major
    * Leading to no release: no-release  
    
The labels get assigned based on our branch naming convention.  
*-patch, *-minor, *-major lead to patch, minor, major label
*-tests, *-documentation, *-configuration, *-force lead to no-release label  

### create_release.yml (Create Release)

Whenever a pull request is closed this action gets triggered.  
Additionally it has a gatekeeper by the check if the pull request was merged and the head_ref started with RC  

It contains the following **Actions**:
* checkout (provided by github actions)
  * checks out the current master  
* Build assets
  * TODO - Build package action has to be created
  * Should have params for all specific shop extensions/repos so 1 action can be used for all extensions/repos  
* Verify changelog
  * TODO - Verify changelog action has to be created
  * Should verify if the semver next tag (or one of them so it doesn't need to check current tag) is in the CHANGELOG.md  
    Check github-create-release-action on how to fetch CHANGELOG.md and look for h2 (##TAG)
* Tag and prepare release
  * uses K-Phoen/semver-release-action
  * Bumps the tag and creates an empty release depending on the label the merged PR had  
    (semver compliant - patch, minor, major)
* Upload release notes and package
  * uses serucee/github-create-release-action
  * Updates the previously created release (passed as param)
    * Adds release notes

## TODO

### Create script for release package creation
* Pass params as env for every different shop system
* Script contains switch case for all shop systems and decides what to add to the release package
* Outputs release package filename (what has to be uploaded) and the displayed filename for the release

### Github settings
* Add label_pull_request.yml as mandatory check before merge can happen

CHANGELOG.md
# Changelog

## v7.0.0

* Adjust tagging workflow
* More release notes

|  Overview | test version | PHP 5.6 | PHP 7.0 | PHP 7.1 | PHP 7.2 |  
|---|---|:---:|:---:|:---:|:---:|  
| **Tested version(s)** | TEST sdg | &#9989; | &#9989; | &#9989; | &#9989; |  
| **Compatibility** | test asd | &#9989; | &#9989; | &#9989; | &#9989; |  

## v6.6.0

* Test release. Title missing.
* Release note
* Another test
* Bump

|  Overview | test version | PHP 5.6 | PHP 7.0 | PHP 7.1 | PHP 7.2 |  
|---|---|:---:|:---:|:---:|:---:|  
| **Tested version(s)** | TEST sdg | &#9989; | &#9989; | &#9989; | &#9989; |  
| **Compatibility** | test asd | &#9989; | &#9989; | &#9989; | &#9989; |  

## v6.5.0

* Testing annotated tag

|  Overview | test version | PHP 5.6 | PHP 7.0 | PHP 7.1 | PHP 7.2 |  
|---|---|:---:|:---:|:---:|:---:|  
| **Tested version(s)** | TEST sdg | &#9989; | &#9989; | &#9989; | &#9989; |  
| **Compatibility** | test asd | &#9989; | &#9989; | &#9989; | &#9989; |  

## v6.4.0

* Testing lightweight tag
* More release notes

|  Overview | test version | PHP 5.6 | PHP 7.0 | PHP 7.1 | PHP 7.2 |  
|---|---|:---:|:---:|:---:|:---:|  
| **Tested version(s)** | TEST sdg | &#9989; | &#9989; | &#9989; | &#9989; |  
| **Compatibility** | test asd | &#9989; | &#9989; | &#9989; | &#9989; |  

## v6.3.1

* Testing lightweight tag
* More release notes

|  Overview | test version | PHP 5.6 | PHP 7.0 | PHP 7.1 | PHP 7.2 |  
|---|---|:---:|:---:|:---:|:---:|  
| **Tested version(s)** | TEST sdg | &#9989; | &#9989; | &#9989; | &#9989; |  
| **Compatibility** | test asd | &#9989; | &#9989; | &#9989; | &#9989; |  

## v6.3.0

* Test
* More release notes

|  Overview | test version | PHP 5.6 | PHP 7.0 | PHP 7.1 | PHP 7.2 |  
|---|---|:---:|:---:|:---:|:---:|  
| **Tested version(s)** | TEST sdg | &#9989; | &#9989; | &#9989; | &#9989; |  
| **Compatibility** | test asd | &#9989; | &#9989; | &#9989; | &#9989; |  

## v6.2.0

* Test release package building
* Test update

## v6.1.1

* Improve error messages in release creation

## v6.1.0

* Test blacklist for files and folders
* Test blacklists

## v6.0.2

* Update release build action
* Retest release
* Retest package build fix

## v6.0.1

* Update release action

## v6.0.0

* Add release package creation

## v5.0.1

* Bump version to trigger release

## v5.0.0

* Do stuff and bump changelog
* Test me

## v4.0.2

* Add release package upload

## v4.0.1

* Test

## v4.0.0

* Added pr-labeler
* Some more breaking changes
* Yes

## v3.0.1

Bump v3.0.0-major test changelog

## v3.0.0

Bump v3.0.0-major test changelog

## v2.0.2

Bump v2.0.2-patch test changelog

## v2.0.1

Bump v2.0.1-patch test changelog

## v2.0.0

Bump major

## v1.0.4

Test workflow 1.0.4

## v1.0.3

Test workflow 1.0.3

## v1.0.2

Test release cause everything works

## v1.0.1

Test release cause everything works

## v1.0.0

Major release cause everything works

## v0.0.5

Test changed workflow trigger

## v0.0.4

Test changed workflow trigger

## v0.0.3

Test new workflow

## v0.0.2

Test pull request merged  

## v0.0.1

Initial Release

.releaserc.js
function camelCaseToEnglish(string) {
    return string.replace( /([A-Z])/g, " $1" );
}

function addNormalizedHeader(context, options) {
    context.commitGroups = context.commitGroups.map(group => {
        group.title = camelCaseToEnglish(group.title);

        return group
    });

    return context
}

function resolveVersionNumbers() {
    const versions = [
        { label: "PHP", compatibility: "5.6, 7.0, 7.1, 7.2", tested: "7.2" },
    ];

    return versions
}

module.exports = {
    "plugins": [
        ["@semantic-release/commit-analyzer", {
            "preset": "eslint",
            "releaseRules": [
                { "tag": "Breaking", "release": "major" },
                { "tag": "PaymentMethods", "release": "minor" },
                { "tag": "Feature", "release": "minor" },
                { "tag": "Fix", "release": "patch" }
            ]
        }],

        ["@semantic-release/release-notes-generator", {
            "preset": "eslint",
            "writerOpts": {
                "commitPartial": "* {{ message }}\n",
                "finalizeContext": addNormalizedHeader
            }
        }],

        ["@semantic-release/github", {
            "assets": [
                { "path": "package.zip", "label": "Package" }
            ]
        }],

        // ["semantic-release-compatibility-table", {
        //     resolveVersionNumbers
        // }]
    ]
};